tc/core/tc_executor_api_rfc.h

```diff
@@ -0,0 +1,145 @@
+ *   using CacheType = CudaCache;
+ *   using CacheEntryType = CudaCacheEntry;
+ *   using CacheKeyType = typename CacheEntryType::Key;
+ *   using CompilationResultType = typename CacheEntryType::Values;
+ *   using RTCFunctionType = CudaRTCFunction;
+ * };
+ *
+ * The correspondence is 1 TcExecutor for 1
+ *   tuple<TC function, input shapes, Backend::MappingOptions>.
+ *
+ * Backend-specific Executors should specialize TcExecutor<Backend> and
+ * implement the uncheckedRun and setRuntimeCompiledFunction.
+ *
+ * Specializing a TcExecutor<Backend> consists in overriding 2 types of methods:
+ * 1. uncheckedRun/profileUnchecked which provides the low-latency path on
+ *    void* data
+ * 2. setupRuntimeCompiledFunction/clearRuntimeCompiledFunction which takes
+ *    the results of compileWithTcMapper to:
+ *       a. save backend-specific information required at runtime (e.g. grid and
+ *          block sizes for the CUDA backend)
+ *       b. create the in-memory object resulting from JIT-compilation that
+ *          implements the TC
+ */
+template <typename Backend>
+class TcExecutor {
+ protected:
+  TcExecutor(
+      lang::TreeRef tcDefinition,
+      const std::vector<const DLConstTensor*>& inputs);
+
+ public:
+  virtual ~TcExecutor();
+
+  TcExecutor(TcExecutor&&) = delete;
+  TcExecutor& operator=(TcExecutor&&) = delete;
+  TcExecutor(const TcExecutor&) = delete;
+  TcExecutor& operator=(const TcExecutor&) = delete;
+
+  /// Run can be called multiple times given a compilation, inputs are allowed
+  /// to change in that their data pointer is allowed to change.
+  /// Sizes and strides must remain constant otherwise this is an error.
+  /// The only thing that is allowed to change across runs is the input
+  /// and output pointers base address.
+  /// It is the caller's responsibility to ensure proper non-aliasing (or
+  /// advanced aliasing) properties of the input and output tensors.
+  void run(
+      const std::vector<const DLConstTensor*>& inputs,
+      const std::vector<const DLTensor*>& outputs) const;
+
+  /// Calls run and profiles the kernel runtime (microseconds).
+  /// \returns profiling information (see: tc/core/utils/time.h)
+  ProfilingInfo profile(
+      const std::vector<const DLConstTensor*>& inputs,
+      const std::vector<const DLTensor*>& outputs) const;
+
+  /*****************************************************************************
+   * The following pure virtual methods are the minimal set of required
+   * backend-dependent functions.
+   ****************************************************************************/
+  /// This is the "low-latency" mode in which we just propagate raw pointers to
+  /// data in the address space where kernel is executed.
+  /// No tensor-related information can be checked so it is the user's
+  /// responsibility to ensure that shapes and strides match. If the user
+  /// doesn't then segfault will likely occur.
+  virtual void uncheckedRun(
```

Comment:
Why the virtual interface? Are specialized `TcExecutor`s (e.g., `TcExecutor<CudaBackend>`) supposed to be inherited from?

Reply:
I had had the same question and came up with the following explanation (but it would be much better if it was documented).
We need inheritance to share certain functionality between executors (non-virtual functions, parameters, etc.).  But we want individual executors to be statically type-safe, e.g. take only the mapping options of the specified type, which would not be possible without templates. So we have _both_ templates and inheritance: `class DefaultCudaExecutor : public TcExecutor<Cuda>`.  As a side effect, it provides a possibility for _multiple_ concrete executors for the same backend.  For example, `class DLTensorCudaExecutor : public TcExecutor<Cuda>` and `class ATenCudaExecutor : public TcExecutor<Cuda>`.

 Another solution here is to use CRTP for static polymorphism, but it would require maintaining a list of inheriting classes and manually dispatching polymorphic functions, which is arguably annoying.

Reply:
I don't think we need virtuals or CRTP. E.g. I'd imagine that `run` does something like:
```
void run(const std::vector<const DLConstTensor*>& inputs,
              const std::vector<const DLTensor*>& outputs) const {
  //check sizes
  uncheckedRun(inputs, outputs)
}
```
`run` is shared among executors and each concrete has to implement `unchechedRun`.
If an executor omits it, then it won't compile

Reply:
Unless we want to use subtypes of `TcExecutors<ConcreteBackend>` polymorphically then there's no need for the virtuals.

Reply:
How would it call `uncheckedRun` if it is not declared (even as pure virtual)?  If it is declared but not virtual, `run` will call `uncheckedRun` from the base class.

Reply:
One way to do it:
```
template <typename T> struct A {
    template <typename U> struct URunner { static void uncheckedRun(); };
    void run() { URunner<T>::uncheckedRun(); }
};

template <> template <> struct A<int>::URunner<int> {

    static void uncheckedRun() {}
};

int main() {
    A<int> a;
    a.run();
    A<char> b;
    b.run(); //causes linker error, probably with some TMP magic it can be turned to a compile error
}
```

Reply:
`static_assert` doesn't work directly in this example, it will trigger regardless of any specializations' existence. 

I don't have  problem with virtual member function but:
A class with a virtual API declares certain invariants/intents one of which would be that it is meant to be used polymorphically. However, according to the documentation we are supposed to specialize the `TcExecutor` and not inherit from it, on the other hand a virtual destructor indicates that we are actually expecting to handle subtypes of `TcExecutor<XBacked>` through pointers to `TcExecutor<XBacked>`.  I find this confusing.

But the use cases you mentioned, e.g. having an `ATen` based executor,  are not handled well with by virtual functions since would have to write wrappers with `at::Tensor` parameters.

Reply:
`Too much template magic for my taste, and having to run double-template specializations is not much better than maintaining CRTP dispatch.`

`uncheckedRun` and `profileUnchecked` could be factored out to a separate struct that is passed as a template argument, e.g.:

```
template <typename Backend, template <typename> typename TcExecutorUnchecked >
class TcExecutor {
//...
void run(const std::vector<const DLConstTensor*>& inputs,
              const std::vector<const DLTensor*>& outputs) const {
  //check sizes
  TcExecutorUnchecked<Backend>::uncheckedRun(*this, inputs, outputs);
}
//...
};
```
And the user of `TcExecutor` has to just create a type `TcExecutorUnchecked` that implements the two functions (or we provide it and it has to be explicitly specialized).

Reply:
I already commented that `Backend` struct seems artificial.  Couldn't we just have a `CudaTcExecutor` (and `ATenCudaTcExecutor`) with all relevant defines.

> `TcExecutorUnchecked<Backend>::uncheckedRun`

This and previous examples assume `uncheckedRun` is static.  So concrete executors cannot have state.  This looks problematic because they usually need state, like grid configuration.

Reply:
> doesn't work directly in this example,

then just don't declare the function in the non-specialized case

>  However, according to the documentation

This is the root of the problem. *The documentation has not been updated* so I stopped reading it on these PRs.

> A class with a virtual API declares certain invariants/intents one of which would be that it is meant to be used polymorphically.

Dynamic dispatch to specific implementations from a shared generic one seems a legit polymorphism usecase for me.

Reply:
> This and previous examples assume `uncheckedRun` is static. So concrete executors cannot have state. This looks problematic because they usually need state, like grid configuration.

In the example I'm passing the `*this` pointer to `uncheckedRun` for this exact reason.

Also I think there's a terminology mismatch which sent me down the template rabbit hole:
the documentation mentions specialization but it should have been (in C++ lingo) instantiation. 
Specializing the whole template class makes little sense since we'd be throwing away all default and
shared implementations.

Reply:
> specialization but it should have been (in C++ lingo) instantiation.
> Specializing the whole template class makes little sense since we'd be throwing away all default and
shared implementations.

Indeed.